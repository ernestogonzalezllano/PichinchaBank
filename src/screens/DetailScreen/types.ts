import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';

export type DetailScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Detail'
>;
