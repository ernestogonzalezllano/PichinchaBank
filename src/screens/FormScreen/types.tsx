import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';
import {GapsNames} from '../../types/api';

export type FormScreenProps = NativeStackScreenProps<
  RootStackParamList,
  'Form'
>;

export type FormGaps = {
  [key in GapsNames]: string | undefined;
};
