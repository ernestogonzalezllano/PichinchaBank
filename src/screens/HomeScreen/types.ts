import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {RootStackParamList} from '../../types/screensTypes';

export type HomeProps = NativeStackScreenProps<RootStackParamList, 'Home'>;
