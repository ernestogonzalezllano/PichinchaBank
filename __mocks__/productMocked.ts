export const productMocked = {
  id: '12345',
  date_release: '2024-11-11',
  date_revision: '2025-11-11',
  description: 'test react testing library',
  logo: 'https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg',
  name: 'test test',
};
