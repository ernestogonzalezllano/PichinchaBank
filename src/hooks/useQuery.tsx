import {useCallback, useEffect, useState} from 'react';
import {BASE_URL} from '../constants';
import {ApiRoutes} from '../types/api';
import {useFocusEffect} from '@react-navigation/native';

interface Props {
  url: ApiRoutes;
  refetchOnMount?: boolean;
  options?: {
    body?: {[key: string]: any};
    method?: 'POST' | 'PUT' | 'DELETE' | 'GET';
    params?: {[key: string]: any};
  };
}

export const useQuery = <T,>({url, refetchOnMount = true, options}: Props) => {
  const [data, setData] = useState<T>();
  const [error, setError] = useState<string>();
  const [loading, setLoading] = useState<boolean>(refetchOnMount ?? false);

  const refetch = useCallback(() => {
    setData(undefined);
    setError(undefined);
    setLoading(true);
    const body = options?.body ? JSON.stringify(options.body) : undefined;
    const params = options?.params
      ? '?' + new URLSearchParams(options.params).toString()
      : '';

    fetch(`${BASE_URL}${url}${params}`, {
      ...options,
      body,
      headers: {
        authorId: '10',
        'Content-Type': 'application/json',
      },
    })
      .then(response => {
        if (!response.ok) {
          throw new Error('Something was wrong');
        }
        return response.text().then(textResponse => {
          let result = textResponse;
          try {
            result = JSON.parse(textResponse);
          } catch {}
          return result as T;
        });
      })
      .then(result => {
        setData(result);
      })
      .catch(() => {
        setError('Error al obtener datos de la API');
      })
      .finally(() => {
        setLoading(false);
      });
  }, [options, url]);

  const myFunction = useCallback(() => {
    if (url && refetchOnMount) {
      refetch();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    return () => {
      setData(undefined);
      setError(undefined);
    };
  }, []);

  useFocusEffect(myFunction);
  return {data, error, refetch, loading};
};
