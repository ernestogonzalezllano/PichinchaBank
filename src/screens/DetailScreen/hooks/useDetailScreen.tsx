import {useState} from 'react';
import {useQuery} from '../../../hooks/useQuery';
import {Product} from '../../../types/api';

export const useDetailScreen = ({product}: {product: Readonly<Product>}) => {
  const [openModal, setOpenModal] = useState(false);
  const {refetch, loading, error, data} = useQuery<string>({
    url: '/bp/products',
    options: {method: 'DELETE', params: {id: product.id}},
    refetchOnMount: false,
  });
  return {openModal, setOpenModal, refetch, loading, error, data};
};
