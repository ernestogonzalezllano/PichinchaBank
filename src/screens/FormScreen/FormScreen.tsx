import React from 'react';
import {StyleSheet, View} from 'react-native';

import {Input} from '../../components/Input';
import {ButtonTouchable} from '../../components/ButtonTouchable';
import {isValidDateString} from '../../utils/validators';
import {GapsNames} from '../../types/api';
import {GAPS} from '../../constants';
import {FormScreenProps} from './types';
import {Modal} from './components/Modal';
import {useFormScreen} from './hooks/useFormScreen';

export const FormScreen = ({
  navigation: {popToTop},
  route: {params: product},
}: FormScreenProps) => {
  const {
    data,
    error,
    errors,
    handleSubmit,
    loading,
    setFormData,
    formData,
    setErrors,
    defaultFormData,
  } = useFormScreen({product});

  return (
    <View style={styles.container}>
      <View>
        {(Object.keys(GAPS) as GapsNames[]).map(gap => (
          <Input
            key={gap}
            {...GAPS[gap]}
            errorLabel={errors[gap as keyof typeof errors]}
            value={formData[gap]}
            onChange={text => {
              const value = {[gap]: text};
              if (gap === 'date_release') {
                const isDate = isValidDateString(text);
                if (isDate) {
                  const revisionDate = new Date(isDate);
                  revisionDate.setFullYear(revisionDate.getFullYear() + 1);
                  value.date_revision = isValidDateString(
                    revisionDate.toString(),
                  ) as string;
                }
              }
              setFormData(oldState => ({
                ...oldState,
                ...value,
              }));
            }}
          />
        ))}
      </View>

      <View style={styles.buttonsContainer}>
        <ButtonTouchable
          loading={loading}
          label="Enviar"
          onPress={handleSubmit}
        />
        <ButtonTouchable
          loading={loading}
          label="Reiniciar"
          type="secondary"
          onPress={() => {
            setFormData(defaultFormData);
            setErrors({});
          }}
        />
        {(data || error) && <Modal {...{popToTop, error}} />}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
  },
  buttonsContainer: {
    width: '100%',
    gap: 10,
  },
});
