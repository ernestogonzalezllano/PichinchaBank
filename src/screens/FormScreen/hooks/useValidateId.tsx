import {useEffect} from 'react';
import {useQuery} from '../../../hooks/useQuery';

export const useValidateId = ({id, isEdit}: {id?: string; isEdit: boolean}) => {
  const {data: validateId, refetch: validateIdRefetch} = useQuery<boolean>({
    url: '/bp/products/verification',
    refetchOnMount: false,
    options: {params: {id}, method: 'GET'},
  });

  useEffect(() => {
    if (id && !isEdit) {
      validateIdRefetch();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [id]);
  return {validateId};
};
