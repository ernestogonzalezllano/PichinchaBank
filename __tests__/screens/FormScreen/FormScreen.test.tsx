import * as React from 'react';
import * as useQuery from '../../../src/hooks/useQuery';
import * as useValidateId from '../../../src/screens/FormScreen/hooks/useValidateId';

import {
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';
import {FormScreen} from '../../../src/screens/FormScreen/FormScreen';
import {FormScreenProps} from '../../../src/screens/FormScreen/types';
import {productMocked as originalProductMocked} from '../../../__mocks__/productMocked';

const productMocked = {...originalProductMocked};
const popToTop = jest.fn();

const props = {
  navigation: {
    popToTop,
  },
  route: {},
} as unknown as FormScreenProps;

describe('FormScreen', () => {
  test('Should match with snapshot', async () => {
    const component = render(
      <NavigationContainer>
        <FormScreen {...props} />
      </NavigationContainer>,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('Should autofill date_revision when date_release is changed', async () => {
    const modifiedProps = {
      ...props,
      route: {...props.route, params: productMocked},
    };
    render(
      <NavigationContainer>
        <FormScreen {...modifiedProps} />
      </NavigationContainer>,
    );
    const input = await screen.findByDisplayValue(productMocked.date_release);
    fireEvent.changeText(input, '2022-11-11');
    await waitFor(() => {
      const newTextElement = screen.getByDisplayValue('2023-11-09');
      expect(newTextElement).toBeOnTheScreen();
    });
  });
  test('Should clean gaps when "Reiniciar" button is pressed', async () => {
    const modifiedProps = {
      ...props,
      route: {...props.route, params: productMocked},
    };
    render(
      <NavigationContainer>
        <FormScreen {...modifiedProps} />
      </NavigationContainer>,
    );
    const button = screen.getByText('Reiniciar');
    fireEvent.press(button);
    await waitFor(() => {
      const algo = screen.queryByDisplayValue(productMocked.date_release);
      expect(algo).toBeNull();
    });
  });
  test('Should validate gaps on submit and show errors', async () => {
    render(
      <NavigationContainer>
        <FormScreen {...props} />
      </NavigationContainer>,
    );
    fireEvent.press(screen.getByText('Enviar'));
    expect(screen.getAllByText('Este campo es requerido')).toHaveLength(5);
  });
  test('Should validate gaps on submit and refetch data', async () => {
    refetchTest();
  });
  test('Should show modal on refetch success', async () => {
    refetchTest();
    await waitFor(() => {
      expect(screen.getByText('Realizado!')).toBeOnTheScreen();
    });
  });
  test('Should go to main page when "Volver al inicio" button is pressed', async () => {
    refetchTest();

    fireEvent.press(screen.getByText('Volver al inicio'));
    expect(popToTop).toHaveBeenCalled();
  });
});

const refetchTest = () => {
  const refetch = jest.fn();

  jest.spyOn(useValidateId, 'useValidateId').mockReturnValue({
    validateId: false,
  });
  jest.spyOn(useQuery, 'useQuery').mockReturnValue({
    data: {data: 'test'},
    error: undefined,
    refetch,
    loading: false,
  });
  const modifiedProps = {
    ...props,
    route: {...props.route, params: productMocked},
  };
  render(
    <NavigationContainer>
      <FormScreen {...modifiedProps} />
    </NavigationContainer>,
  );
  fireEvent.press(screen.getByText('Enviar'));
  expect(refetch).toHaveBeenCalled();
};
