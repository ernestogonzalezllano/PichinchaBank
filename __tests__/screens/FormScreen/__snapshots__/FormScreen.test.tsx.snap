// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormScreen Should match with snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View>
    <View>
      <Text
        style={
          {
            "fontWeight": "bold",
            "marginVertical": 5,
          }
        }
      >
        ID
      </Text>
      <TextInput
        editable={true}
        onChangeText={[Function]}
        style={
          {
            "borderColor": "#eaecf3",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          }
        }
      />
    </View>
    <View>
      <Text
        style={
          {
            "fontWeight": "bold",
            "marginVertical": 5,
          }
        }
      >
        Nombre
      </Text>
      <TextInput
        editable={true}
        onChangeText={[Function]}
        style={
          {
            "borderColor": "#eaecf3",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          }
        }
      />
    </View>
    <View>
      <Text
        style={
          {
            "fontWeight": "bold",
            "marginVertical": 5,
          }
        }
      >
        Descripción
      </Text>
      <TextInput
        editable={true}
        onChangeText={[Function]}
        style={
          {
            "borderColor": "#eaecf3",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          }
        }
      />
    </View>
    <View>
      <Text
        style={
          {
            "fontWeight": "bold",
            "marginVertical": 5,
          }
        }
      >
        Logo
      </Text>
      <TextInput
        editable={true}
        onChangeText={[Function]}
        style={
          {
            "borderColor": "#eaecf3",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          }
        }
        value="https://www.visa.com.ec/dam/VCOM/regional/lac/SPA/Default/Pay%20With%20Visa/Tarjetas/visa-signature-400x225.jpg"
      />
    </View>
    <View>
      <Text
        style={
          {
            "fontWeight": "bold",
            "marginVertical": 5,
          }
        }
      >
        Fecha liberación
      </Text>
      <TextInput
        editable={true}
        onChangeText={[Function]}
        style={
          {
            "borderColor": "#eaecf3",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          }
        }
      />
    </View>
    <View>
      <Text
        style={
          {
            "fontWeight": "bold",
            "marginVertical": 5,
          }
        }
      >
        Fecha revisión
      </Text>
      <TextInput
        editable={false}
        onChangeText={[Function]}
        style={
          {
            "backgroundColor": "#eaecf3",
            "borderColor": "#eaecf3",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "gap": 10,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#ffdd04",
          "borderRadius": 5,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "backgroundColor": "#ffdd04",
            "fontWeight": "bold",
          }
        }
      >
        Enviar
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#eaecf3",
          "borderRadius": 5,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "backgroundColor": "#eaecf3",
            "fontWeight": "bold",
          }
        }
      >
        Reiniciar
      </Text>
    </View>
  </View>
</View>
`;
